src/coreclr/jit/valuenum.cpp

```diff
@@ -3675,7 +3675,10 @@ ValueNum ValueNumStore::VNForMapSelectWork(ValueNumKind      vnk,
                     entry.Result = sameSelResult;
                     entry.SetMemoryDependencies(m_pComp, recMemoryDependencies);
 
-                    GetMapSelectWorkCache()->Set(fstruct, entry);
+                    // If we ran out of budget we could have already cached the
+                    // result in the leaf. In that case it is ok to overwrite
+                    // it here.
+                    GetMapSelectWorkCache()->Set(fstruct, entry, MapSelectWorkCache::Overwrite);
```

Comment:
Is it worth validating in debug that the cache has this result already?

Reply:
Added an assert for this